### PR TITLE
feat(pipeline): support dispatch cron to edge when edge side pipeline registered

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
support dispatch cron to edge when edge side pipeline registered

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=313905&iterationID=-1&pId=0&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   support dispatch cron to edge when edge side pipeline registered           |
| 🇨🇳 中文    |    dge pipeline 注册上来之后，center pipeline 主动下发相关 cron          |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: apistructs/cluster_dialer.go [hunk 1]
```diff
@@ -45,6 +45,10 @@ func (c ClusterManagerClientType) String() string {
 	return string(c)
 }
 
+func (c ClusterManagerClientType) GenEventName() string {
```
Comment (sfwn) [inline apistructs/cluster_dialer.go:48]: I think this is just the big scope of event.
for one client type, you maybe have multiple event name.

File: bundle/cluster_manager.go [hunk 2]
```diff
@@ -65,3 +65,24 @@ func (b *Bundle) GetClusterManagerClientData(clientType apistructs.ClusterManage
 	}
 	return getResp, nil
 }
+
+func (b *Bundle) ListClusterManagerClientByType(clientType apistructs.ClusterManagerClientType) ([]apistructs.ClusterManagerClientDetail, error) {
```
Comment (sfwn) [inline bundle/cluster_manager.go:69]: what is ClusterManagerClient

File: internal/core/cluster-manager/dialer/server/tunnel-server.go [hunk 3]
```diff
@@ -370,12 +384,13 @@ func getClusterClient(server *remotedialer.Server, clusterKey string, timeout ti
...
 }
 
-func NewDialerRouter(ctx context.Context, clusterSvc clusterpb.ClusterServiceServer, credential tokenpb.TokenServiceServer, cfg *conf.Conf, etcd *clientv3.Client) *mux.Router {
+func NewDialerRouter(ctx context.Context, clusterSvc clusterpb.ClusterServiceServer, credential tokenpb.TokenServiceServer, cfg *conf.Conf, etcd *clientv3.Client, bdl *bundle.Bundle) *mux.Router {
 	authorizer := auth.New(
 		auth.WithCredentialClient(credential),
 		auth.WithConfig(cfg),
 	)
 
+	InitClientData(bdl)
```
Comment (sfwn) [inline internal/core/cluster-manager/dialer/server/tunnel-server.go:393]: a little strange

File: internal/tools/pipeline/providers/cron/cron.service.go [hunk 4]
```diff
@@ -490,13 +491,58 @@ func (s *provider) disable(cron *db2.PipelineCron, option mysqlxorm.SessionOptio
...
+		p.Log.Warnf("cluster name is empty， ignore event")
+		return
+	}
+	edgeCrons, err := p.CronPaging(ctx, &pb.CronPagingRequest{
+		ClusterName: clusterName,
+		GetAll:      true,
+		AllSources:  true,
+	})
+	if err != nil {
+		p.Log.Errorf("failed to get edge pipeline crons, clusterName: %v error %v", clusterName, err)
```
Comment (sfwn) [inline internal/tools/pipeline/providers/cron/cron.service.go:507]: , errror

File: internal/tools/pipeline/providers/cron/cron.service.go [hunk 5]
```diff
@@ -490,13 +491,58 @@ func (s *provider) disable(cron *db2.PipelineCron, option mysqlxorm.SessionOptio
...
+			ConfigManageNamespaces: cron.Extra.ConfigManageNamespaces,
+			CronStartFrom:          cron.CronStartTime,
+			IncomingSecrets:        cron.Extra.IncomingSecrets,
+			PipelineDefinitionID:   cron.PipelineDefinitionID,
+		})
+		if err != nil {
+			p.Log.Errorf("failed to create edge pipeline cron, clusterName: %v error %v", clusterName, err)
+			continue
+		}
+		p.Log.Infof("create edge pipeline cron %v success", cron.ID)
```
Comment (sfwn) [inline internal/tools/pipeline/providers/cron/cron.service.go:535]: clearly.
create edge pipeline cron success, cronID: %d

File: internal/tools/pipeline/providers/edgepipeline_register/edge_client.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/mohae/deepcopy"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/pkg/discover"
+	"github.com/erda-project/erda/pkg/http/httpserver"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+var (
+	ClientsHookApiPath = "/api/pipeline-clients/actions/hook"
```
Comment (sfwn) [inline internal/tools/pipeline/providers/edgepipeline_register/edge_client.go:33]: what is pipeline client

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/cluster_dialer.go (hunk 1 above)
Comment (chengjoey) [reply apistructs/cluster_dialer.go:48]: adjusted, add event type

File: bundle/cluster_manager.go (hunk 2 above)
Comment (chengjoey) [reply bundle/cluster_manager.go:69]: image

clusterDialerClient -> clusterManagerClient
clusterDialer was uniformly named clusterManager

File: internal/tools/pipeline/providers/edgepipeline_register/edge_client.go (hunk 6 above)
Comment (chengjoey) [reply internal/tools/pipeline/providers/edgepipeline_register/edge_client.go:33]: adjusted, EdgeHookApiPath = "/api/pipeline-edge/actions/hook"

File: internal/tools/pipeline/providers/cron/cron.service.go (hunk 5 above)
Comment (chengjoey) [reply internal/tools/pipeline/providers/cron/cron.service.go:535]: adjusted

File: internal/tools/pipeline/providers/cron/cron.service.go (hunk 4 above)
Comment (chengjoey) [reply internal/tools/pipeline/providers/cron/cron.service.go:507]: adjusted

File: internal/core/cluster-manager/dialer/server/tunnel-server.go (hunk 3 above)
Comment (chengjoey) [reply internal/core/cluster-manager/dialer/server/tunnel-server.go:393]: any good idea...

File: internal/core/cluster-manager/dialer/server/client_data.go [hunk 7]
```diff
@@ -16,16 +16,26 @@ package server
...
 )
 
 var (
 	clientMutex sync.Mutex
 	clientDatas = map[apistructs.ClusterManagerClientType]apistructs.ClusterManagerClientMap{}
+	bdl         *bundle.Bundle
 )
 
-func updateClientDetail(clientType apistructs.ClusterManagerClientType, clusterKey string, data apistructs.ClusterManagerClientDetail) {
+func InitClientData(b *bundle.Bundle) {
```
Comment (sfwn) [inline internal/core/cluster-manager/dialer/server/client_data.go:34]: no need export it

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/tools/pipeline/providers/edgepipeline_register/edge_client.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/mohae/deepcopy"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/pkg/discover"
+	"github.com/erda-project/erda/pkg/http/httpserver"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+var (
+	EdgeHookApiPath = "/api/pipeline-edge/actions/hook"
```
[the remark this answers is not in the document]
Comment (chengjoey) [reply internal/tools/pipeline/providers/edgepipeline_register/edge_client.go:34]: added

File: internal/core/cluster-manager/dialer/server/client_data.go (hunk 7 above)
Comment (chengjoey) [reply internal/core/cluster-manager/dialer/server/client_data.go]: adjusted

**Review by sfwn — APPROVED** [review]

**Pull request merged**